cmd/loadTests.go

```diff
@@ -703,7 +785,162 @@ func userJourneyThread(frameworkMap *sync.Map, threadWaitGroup *sync.WaitGroup,
+						increaseBar(deploymentsBar, deploymentsBarMutex)
+					}
+					return deploymentIsDone || deploymentFailed, nil
+				})
+				if err != nil {
+					logError(18, fmt.Sprintf("Deployment for %s/%s failed to succeed within %v: %v", applicationName, componentName, deploymentTimeout, err))
+					FailedDeploymentsPerThread[threadIndex] += 1
+					increaseBar(deploymentsBar, deploymentsBarMutex)
+					continue
+				}
+			}
 		}()
 	}
 	wg.Wait()
 }
+
+func checkDeploymentFailed(deployment *appsv1.Deployment) (bool, string, metav1.Time) {
+	var lastTransitionTime metav1.Time = metav1.Now() // initialize with the current time
+
+	// Check if the Deployment is in a stable state
+	if deployment.Status.ObservedGeneration < deployment.Generation {
+		return false, "", lastTransitionTime
+	}
+
+	// Check if the Deployment is unable to progress
+	for _, condition := range deployment.Status.Conditions {
+		if condition.Type == appsv1.DeploymentProgressing && condition.Status == corev1.ConditionFalse {
+			lastTransitionTime = condition.LastTransitionTime
+			return true, fmt.Sprintf("Deployment failed to progress: %s", condition.Message), lastTransitionTime
+		}
+	}
+
+	// Check if the Deployment couldn't create the required number of pods
+	if deployment.Spec.Replicas != nil && deployment.Status.AvailableReplicas < *deployment.Spec.Replicas {
+		for _, condition := range deployment.Status.Conditions {
+			if condition.Type == appsv1.DeploymentAvailable && condition.Status == corev1.ConditionFalse {
+				lastTransitionTime = condition.LastTransitionTime
+				break
+			}
+		}
+		return true, fmt.Sprintf("Deployment failed to create required pods: wanted %d, have %d", *deployment.Spec.Replicas, deployment.Status.AvailableReplicas), lastTransitionTime
+	}
+
+	// Check for errors during rollout
+	if deployment.Status.UpdatedReplicas < deployment.Status.Replicas {
+		for _, condition := range deployment.Status.Conditions {
+			if condition.Type == appsv1.DeploymentReplicaFailure {
+				lastTransitionTime = condition.LastTransitionTime
+				break
+			}
+		}
+		return true, fmt.Sprintf("Deployment failed during rollout: updated %d, total %d", deployment.Status.UpdatedReplicas, deployment.Status.Replicas), lastTransitionTime
+	}
+
+	return false, "", lastTransitionTime
+}
+
+func checkDeploymentIsDone(deployment *appsv1.Deployment) (bool, metav1.Time) {
+	var deploymentRolloutFinishedUpdatingReplicas bool = false
+	var deploymentReachedDesiredState bool = false
+	var CompletionTime metav1.Time = metav1.Now() // initialize with the current time
+
+	// Check if the Deployment is in a stable state
+	if deployment.Status.ObservedGeneration < deployment.Generation {
+		return false, CompletionTime
+	}
+
+	// Check the Progressing condition
+	for _, condition := range deployment.Status.Conditions {
+		if condition.Type == appsv1.DeploymentProgressing && condition.Status == corev1.ConditionTrue && condition.Reason == "NewReplicaSetAvailable" {
```

Comment:
I detect a deployment as done when it's condition.Type `Available` status is true. I do not have experience with `DeploymentProgressing` condition. Is it OK?

Reply:
The appsv1.DeploymentProgressing condition type signifies that the deployment is in a state of change. The status being corev1.ConditionTrue and reason being "NewReplicaSetAvailable" specifically means that a new ReplicaSet was created and has finished its rollout

The condition type "Available" is also used in deployments, but it signifies a different state. If condition.Type == appsv1.DeploymentAvailable and condition.Status == corev1.ConditionTrue, it means that the minimum required replicas for the deployment are available. This doesn't necessarily mean that the rollout has finished, as there might still be more replicas being created.

Reply:
Nice. Please do you have a link to the docs with this ^^^?

Reply:
One more thing,  The LastTransitionTime from the DeploymentAvailable condition may not always signify the exact moment that the deployment reached its desired state  because the DeploymentAvailable condition could have transitioned to True before all replicas were updated and ready.

Reply:
updated cmd/loadTests.go
fixed checkDeploymentIsDone and checkDeploymentFailed 
verified with documentation

/* Check the DeploymentProgressing condition
   https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#complete-deployment
	
   When the rollout becomes “complete”, the Deployment controller sets a condition with the following attributes to the Deployment's .status.conditions:
   type: Progressing
   status: "True"
   reason: NewReplicaSetAvailable
*/